### DISCUSSION TITLE
enh: open-webui:slim docker image

### DISCUSSION THREAD

**Discussion by tjbck** [open-webui/open-webui#12801] [Ideas]

**Comment by powerman** [comment]
It's 11GB now! This is really weird keeping in mind it's just an UI for an external service…

**Reply by timdinhdotcom** [reply]
It's bigger than a desktop operating system lol

**Reply by powerman** [reply]
Actually that's not funny. I had to stop using open-webui because of this only reason (and partly because they publish docker images on ghcr.io which is really slow, and together with such huge image size this result in hours downloading on each update).

**Reply by rgaricano** [reply]
@powerman 

You can clone the source code and build the Docker image on your side. I don't use Docker, and the downloads are fast. Since I've already downloaded and installed the dependent libraries, the process is quick. As I mentioned, I don't use Docker, but I've built Docker images without any problems.

I would also prefer more lightness, as has been reiterated: proposals are welcome.

**Reply by powerman** [reply]
@rgaricano I'm using kind of IaaC where all services are installed and updated using `docker compose` and dependabot-managed updates. So, building from source means actually building in a docker plus some dependabot-compatible way to detect new versions plus a costly GitHub Actions (that IaaC repo is a private) in case building will take noticeable amount of time (keeping docker cache layers in CI is tricky and with such a huge image will also be costly enough). Honestly, I don't see any usable solution here.

As for proposals - I doubt anyone from outside will be able to find a good way to solve this issue, this task is for core devs. And very likely it'll require fair amount of work to optimize this, which in turn will delay new features. New features are important, no doubt, but they are not an absolute goodness by itself - new features are good because they bring you new users… but this have to be balanced with losing users because of this issue.

**Reply by rgaricano** [reply]
I don't see the optimization aspect as being that problematic either.

You just need to install the libraries used (Python and Node) dynamically and on a separate volume.
Dynamically, because there are several that may never be used (they are mutually exclusive), these can be loaded only when needed.
If we install them on a separate permanent volume, we gain the advantage of only needing to download the updated ones (which isn't common, and can be done with scripts); that volume and/or the updates could also be downloaded as a package if necessary.

**Comment by rare-magma** [comment]
This PR will reduce the docker image size by ~1.2GB: https://github.com/open-webui/open-webui/pull/17052

**Comment by a1cd** [comment]
I think the slim image should move in the direction of offloading all super heavy workloads off the container itself. 
Anything AI related should be removed if it won't be used. So stuff like PyTorch and PyArrow are pretty unnecessary for slim. These are only the biggest contributors to size, there are plenty of deps that are smaller than most that increase the size significantly together. It would be nice to not have to have them if they aren't used at all. 
This was my original idea for the slim image, but I don't know much about what each package does and all the dependencies and when stuff is used.

If anyone knows are these super large packages important to owui functioning when it's offloading everything else?

Treeshaking I think is easier in JS than python so maybe that could be implemented in a large build step, but thats not easy with python.

**Comment by recrudesce** [comment]
Just pulled `:latest` and it's 2.4gb.  Pulled `:latest-slim` and it's 5.2gb.
